snapfaas/src/fs/mod.rs

```diff
@@ -0,0 +1,266 @@
+    File(File)
+}
+
+mod errors {
+    #[derive(Debug)]
+    pub enum LinkError {
+        LabelError(LabelError),
+        Exists
+    }
+
+    #[derive(Debug)]
+    pub enum UnlinkError {
+        LabelError(LabelError),
+        DoesNotExists
+    }
+
+    #[derive(Debug)]
+    pub enum LabelError {
+        CannotRead,
+        CannotWrite,
+    }
+}
+
+impl<S: BackingStore> FS<S> {
+    pub fn root(&self) -> Directory {
+        Directory {
+            label: DCLabel::public(),
+            object_id: 0
+        }
+    }
+
+    pub fn create_directory(&self, label: DCLabel) -> Directory {
+        let dir_contents = serde_json::ser::to_vec(&HashMap::<String, DirEntry>::new()).unwrap_or((&b"{}"[..]).into());
+        let mut uid: UID = rand::random();
+        while !self.storage.add(uid.to_be_bytes(), &dir_contents) {
+            uid = rand::random();
+        }
+
+        Directory {
+            label,
+            object_id: uid,
+        }
+    }
+
+    pub fn create_file(&self, label: DCLabel) -> File {
+        let mut uid: UID = rand::random();
+        while !self.storage.add(uid.to_be_bytes(), &[]) {
+            uid = rand::random();
+        }
+        File {
+            label,
+            object_id: uid,
+        }
+    }
+
+    pub fn list(&self, dir: Directory) -> Result<HashMap<String, DirEntry>, LabelError> {
+        CURRENT_LABEL.with(|current_label| {
+            if dir.label.can_flow_to(&*current_label.borrow()) {
+                Ok(match self.storage.get(dir.object_id.to_be_bytes()) {
+                    Some(bs) => serde_json::from_slice(bs.as_slice()).unwrap_or_default(),
+                    None => Default::default()
+                })
+            } else {
+                Err(LabelError::CannotRead)
+            }
+        })
+    }
+
+    pub fn link(&self, dir: &Directory, name: String, direntry: DirEntry) -> Result<String, LinkError>{
+        CURRENT_LABEL.with(|current_label| {
+            if !current_label.borrow().secrecy.implies(&dir.label.secrecy) {
+                return Err(LinkError::LabelError(LabelError::CannotRead));
+            }
+            if !current_label.borrow().can_flow_to(&dir.label) {
+                return Err(LinkError::LabelError(LabelError::CannotWrite));
+            }
+            let mut raw_dir = self.storage.get(dir.object_id.to_be_bytes());
+            loop {
+                let mut dir_contents: HashMap<String, DirEntry> = raw_dir.as_ref().and_then(|dir_contents| serde_json::from_slice(dir_contents.as_slice()).ok()).unwrap_or_default();
+                if let Some(_) = dir_contents.insert(name.clone(), direntry.clone()) {
+                    return Err(LinkError::Exists)
+                }
+                raw_dir = self.storage.cas(dir.object_id.to_be_bytes(), raw_dir, serde_json::to_vec(&dir_contents).unwrap_or((&b"{}"[..]).into()).into());
+                if raw_dir.is_none() {
+                    return Ok(name)
+                }
+            }
```

Comment:
Why do we loop here?

I don't think this `link` implementation is atomic.
A possible, problematic interleaving can be that `get` succeeds and then someone else delete the directory and then `cas` returns a Some. Due to the loop, we try again but I cannot see how the loop is helping?

Reply:
In general, this is optimistic concurrency control: i.e. we assume eventually the operation will go through without conflicts.

I think one source of confusion is that, because we're dealing with objects, the directory can't be deleted, it can just become empty (until it is garbage collected due to a zero ref count). So an `is_none` result from cas actually means there was no conflict and the write succeeded. If it didn't succeed it means the value changed and the result is the new value, so looping tries again using the new value as the starting point.

Perhaps it would be more clear to use a more explicit return type (e.g. `Result<(), Option<Vec<u8>>`), rather than relying on this invariant

Reply:
But we are not trying with new `raw_dir` since `get` is outside the loop?

Reply:
cas returns the new value if it doesn't match the expected value, so it's like doing a get again after checking

Reply:
Got it.